pkg/cache/cache_test.go

```diff
@@ -0,0 +1,214 @@
+package cache_test
+
+import (
+	"context"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	"github.com/openshift-psap/special-resource-operator/pkg/cache"
+	"github.com/openshift-psap/special-resource-operator/pkg/clients"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func TestCache(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Cache Suite")
+}
+
+func resetGlobals() {
+	clients.Interface = nil
+
+	cache.Node.Count = 0xDEADBEEF
+	cache.Node.List = &unstructured.UnstructuredList{
+		Object: map[string]interface{}{},
+		Items:  []unstructured.Unstructured{},
+	}
+}
+
+var _ = Describe("Cache", func() {
+	Context("Nodes", func() {
+		var (
+			ctrl        *gomock.Controller
+			mockClients *clients.MockClientsInterface
+		)
+
+		BeforeEach(resetGlobals)
+
+		BeforeEach(func() {
+			ctrl = gomock.NewController(GinkgoT())
+			mockClients = clients.NewMockClientsInterface(ctrl)
+
+			clients.Interface = mockClients
+		})
+
+		AfterEach(func() {
+			ctrl.Finish()
+		})
+		AfterEach(resetGlobals)
+
+		listMatcher := gomock.AssignableToTypeOf(&unstructured.UnstructuredList{})
+
+		When("machingLabels=nil", func() {
+			When("force=false", func() {
+				When("no taints", func() {
+					It("should work as expected", func() {
+						k8sItems := make([]unstructured.Unstructured, 2)
+
+						mockClients.EXPECT().List(context.TODO(), listMatcher).Do(
+							func(_ context.Context, l *unstructured.UnstructuredList) {
+								l.Items = k8sItems
+							},
+						)
+
+						err := cache.Nodes(nil, false)
+						Expect(err).NotTo(HaveOccurred())
+						Expect(cache.Node.List.Items).To(HaveLen(len(k8sItems)))
+					})
+				})
+
+				DescribeTable("taints",
+					func(te v1.TaintEffect) {
+						const validNodeName = "node1"
+
+						mockClients.EXPECT().List(context.TODO(), listMatcher).Do(
+							func(_ context.Context, l *unstructured.UnstructuredList) {
+								validNode := unstructured.Unstructured{}
+								validNode.SetName(validNodeName)
+
+								taintedNode := unstructured.Unstructured{
+									Object: map[string]interface{}{
+										"spec": map[string]interface{}{
+											"taints": []interface{}{
+												map[string]interface{}{"effect": string(te)},
+											},
+										},
+									},
+								}
+
+								l.Items = []unstructured.Unstructured{validNode, taintedNode}
+							},
+						)
+
+						err := cache.Nodes(nil, false)
+
+						Expect(err).NotTo(HaveOccurred())
+						Expect(cache.Node.List.Items).To(HaveLen(1))
+						Expect(cache.Node.List.Items[0].GetName()).To(Equal(validNodeName))
+					},
+					Entry("NoExecute", v1.TaintEffectNoExecute),
+					Entry("NoSchedule", v1.TaintEffectNoSchedule),
+				)
+			})
+
+			When("valid cache", func() {
```

Comment:
IMO there is a lot of code duplication here because of all those `When`, for example you have `cache.Node = cache.NodesCache{` multiple times.

I would suggest removing the `When`s and having bigger `It`, you can divide the `It` to logical steps if you want using `By`.

Here is an example: https://github.com/openshift/assisted-service/blob/master/internal/cluster/progress_test.go